rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,370 @@
+  input_ids = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_ids', dtype=tf.int32)
+  input_mask = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_mask', dtype=tf.int32)
+  label_ids = tf.keras.layers.Input(shape=(128,), name='input_type_ids', dtype=tf.int32)
+  x = encoder([input_ids, input_mask, label_ids])['pooled_output']
+  x = tf.keras.layers.Dropout(rate=0.1)(x)
+  output = tf.keras.layers.Dense(1, activation='sigmoid')(x)
+  model = tf.keras.Model(inputs=[input_ids, input_mask, label_ids], outputs=output)
+
+model.compile('adam', 'binary_crossentropy', ['accuracy'])
+model.fit(train_ds, epochs=5, validation_data=test_ds)
+```
+
+### Pretraining task
+
+We aim to provide SavedModel for BertEncoder, however user can choose to pretrain a new BertEncoder based on their
+own dataset.
+
+```python
+with strategy.scope():
+  encoder = keras_nlp.encoders.BertEncoder(vocab_size, max_sequence_length, type_vocab_size)
+  input_ids = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_ids', dtype=tf.int32)
+  input_mask = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_mask', dtype=tf.int32)
+  label_ids = tf.keras.layers.Input(shape=(128,), name='input_type_ids', dtype=tf.int32)
+  masked_lm_positions = tf.keras.layers.Input(shape=(None,), name='masked_lm_positions', dtype=tf.int32)
+  x = encoder([input_ids, input_mask, label_ids])['pooled_output']
+  cls_output, sequence_output = output['pooled_output'], outputs['sequence_output']
+  masked_lm = keras_nlp.layers.MaskedLM(embedding_table=encoder.get_embedding_table())
+  lm_output = masked_lm(sequence_output, masked_positions=masked_lm_positions)
+  cls_output = tf.keras.layers.Dense(units=num_classes, activation='softmax')(cls_output)
+  model = tf.keras.Model(inputs=[input_ids, input_mask, label_ids, masked_lm_positions],
+                         outputs={'lm_output': masked_lm, 'cls_output': cls_output})
+
+model.compile('adam', {'lm_output': 'sparse_categorical_crossentropy', 'cls_output': 'sparse_categorical_crossentropy'})
+model.fit(train_ds, epochs=100)
+```
+
+### Other encoder based network
+
+BertEncoder is the first encoder network we propose in this doc. However other encoder networks can be easily
+built on top of Transformer Encoder layer. For example, for a transformer encoder sharing mechanism
+with [ALBERT](https://arxiv.org/pdf/1909.11942.pdf), this can be achieved by:
+
+```python
+word_ids = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_word_ids')
+mask = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_mask')
+type_ids = tf.keras.layers.Input(shape=(None,), dtype=tf.int32, name='input_type_ids')
+embedding_layer = keras_nlp.layers.OnDeviceEmbedding(vocab_size, embedding_width)
+word_embeddings = embedding_layer(word_ids)
+position_embedding_layer = keras_nlp.layers.PositionEmbedding(max_sequence_length)
+position_embeddings = position_embedding_layer(word_embeddings)
```

Comment:
nit: to be consistent with layers below, maybe we'd also anonymize embedding_layer and position_embedding_layer?
i.e. word_embeddings = keras_nlp.layers.OnDeviceEmbedding(vocab_size, embedding_width)(word_ids)
position_embeddings = keras_nlp.layers.PositionEmbedding(max_sequence_length)(word_embeddings)

Reply:
Done.